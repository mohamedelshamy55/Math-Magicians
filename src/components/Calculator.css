.calc-wrapper {
  display: flex;
  justify-content: space-between;
  align-items: baseline;
  padding: 0 36px;
  width: 100vw;
  height: 75vh;
}

.calc-title {
  width: 40%;
  display: flex;
  margin-top: 36px;
  justify-content: start;
}

.calc-title h3 {
  font-size: 2rem;
  color: thistle;
}

.calcbody {
  border: none;
  box-shadow: inset -3px -3px 5px #ffffff90, 3px 3px 15px #00000070;
  width: 60%;
  height: 496px;
  display: flex;
  flex-direction: column;
  flex-wrap: wrap;
  margin: 12px;
  padding: 0;
  box-sizing: border-box;
  border-radius: 12px;
  overflow: hidden;
}

.display {
  display: flex;
  justify-content: flex-end;
  align-items: center;
  width: 100%;
  height: 95px;
  background: rgb(0, 0, 0);
  background: linear-gradient(111deg, rgba(0, 0, 0, 1) 0%, rgba(51, 51, 51, 1) 52%, rgba(34, 40, 49, 1) 100%);
  box-shadow: inset -3px -3px 5px #ffffff70, inset 3px 3px 15px #00000070;
  text-align: end;
  margin: 0;
  padding-right: 12px;
  overflow: hidden;
}

.result {
  color: #e0e0e0;
  font-size: 2rem;
}

.keypad {
  width: 100%;
  height: 400px;
  display: flex;
}

.numbers {
  background-color: #e0e0e0;
  width: 75%;
  height: 100%;
  display: flex;
  flex-wrap: wrap;
  justify-content: flex-start;
  align-items: flex-start;
  padding: 0;
  margin: 0;
}

.numbers li {
  display: flex;
  justify-content: center;
  align-items: center;
  list-style: none;
  width: 33.3%;
  height: 20%;
  border: none;
  text-align: center;
  margin: 0;
  padding: 0;
}

.aritmetic {
  width: 25%;
  height: 100%;
  background-color: #f5903e;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  padding: 0;
  margin: 0;
}

.aritmetic li {
  display: flex;
  justify-content: center;
  align-items: center;
  list-style: none;
  width: 100%;
  height: 20%;
  border: none;
  text-align: center;
  margin: 0;
  padding: 0;
}

.numbers li button,
.aritmetic li button {
  background-color: transparent;
  font-weight: bold;
  width: 100%;
  height: 100%;
  border: none;
  cursor: pointer;
}

.numbers li button {
  background: #d1d1d1;
  box-shadow: -2px -2px 0 #ffffff70, 3px 3px 10px #00000070;
}

.aritmetic li button {
  font-size: 1.05rem;
  background-color: #f5903e;
  box-shadow: -2px -2px 0 #ffffff70, 3px 3px 15px #00000070;
  color: rgba(0, 0, 0, 0.8);
}

.numbers li button:hover,
.aritmetic li button:hover {
  color: dodgerblue;
}

.numbers li button:active,
.aritmetic li button:active {
  box-shadow: inset -1px -2px 2px #ffffff70, inset 3px 3px 25px #00000070;
  transition: all 170ms ease-in-out;
}

.numbers li[btnName="0"] {
  flex: 2;
}
