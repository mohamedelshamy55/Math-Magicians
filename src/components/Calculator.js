import React, { useState } from 'react';
import './Calculator.css';
import calculate from '../logic/calculate';
import Nums from './Nums';
import Operator from './Operator';
import { keyboard, operators } from './Keyboard';

const calculator = () => {
  const [calcObj, setCalcObject] = useState({});

  const clickHandler = (e) => {
    const buttonName = e.target.innerText;
    setCalcObject(calculate(calcObj, buttonName));
  };

  const screen = `${((calcObj.total || '') + (calcObj.operation || '') + (calcObj.next || '')) || '0'}`;
  return (
    <div data-testid="calc" className="calc-wrapper">
      <div className="calc-title">
        <h3>Let&apos;s do some math!</h3>
      </div>
      <div className="calcbody">
        <div className="display"><p className="result" data-testid="display">{screen}</p></div>
        <div className="keypad">
          <ul className="numbers">
            {keyboard.map((item) => {
              const { id, buttonName } = item;
              return (
                <Nums
                  key={id}
                  buttonName={buttonName}
                  btnname={buttonName}
                  clickHandler={clickHandler}
                />
              );
            })}
          </ul>
          <ul className="aritmetic">
            {operators.map((item) => {
              const { id, buttonName } = item;
              return (
                <Operator
                  key={id}
                  buttonName={buttonName}
                  clickHandler={clickHandler}
                />
              );
            })}
          </ul>
        </div>
      </div>
    </div>
  );
};

export default calculator;
