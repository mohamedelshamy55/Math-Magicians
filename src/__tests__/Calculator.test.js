import React from 'react';
import { fireEvent, render } from '@testing-library/react/';
import '@testing-library/jest-dom/extend-expect';
import Calculator from '../components/Calculator';
import calculate from '../logic/calculate';

const renderer = require('react-test-renderer');

it('matches snapshot 1 of Calculator component', () => {
  const clickHandler = () => true;
  const buttonName = '5';
  const tree = renderer.create(<Calculator
    buttonName={buttonName}
    clickHandler={clickHandler}
  />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders the props of component correctly', () => {
  const clickHandler = () => true;
  const buttonName = 'AC';
  const { getByTestId } = render(<Calculator
    buttonName={buttonName}
    clickHandler={clickHandler}
  />);
  expect(getByTestId('calc')).toHaveTextContent('AC');
});

it('should perform user interaction', () => {
  const { getByRole } = render(<Calculator />);
  const btn = getByRole('button', { name: /=/i });
  const calcObj = {
    total: '3',
    next: '4',
    operation: '+',
  };
  fireEvent.click(btn, { target: { innerText: '=' } });
  const result = calculate(calcObj, btn.value);
  expect(result.total).toBe('7');
});
