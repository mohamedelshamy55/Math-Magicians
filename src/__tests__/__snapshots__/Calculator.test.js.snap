// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1 of Calculator component 1`] = `
<div
  className="calc-wrapper"
  data-testid="calc"
>
  <div
    className="calc-title"
  >
    <h3>
      Let's do some math!
    </h3>
  </div>
  <div
    className="calcbody"
  >
    <div
      className="display"
    >
      <p
        className="result"
        data-testid="display"
      >
        0
      </p>
    </div>
    <div
      className="keypad"
    >
      <ul
        className="numbers"
      >
        <li
          btnname="AC"
          data-testid="num-li"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            AC
          </button>
        </li>
        <li
          btnname="+/-"
          data-testid="num-li"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            +/-
          </button>
        </li>
        <li
          btnname="%"
          data-testid="num-li"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            %
          </button>
        </li>
        <li
          btnname="7"
          data-testid="num-li"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            7
          </button>
        </li>
        <li
          btnname="8"
          data-testid="num-li"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            8
          </button>
        </li>
        <li
          btnname="9"
          data-testid="num-li"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            9
          </button>
        </li>
        <li
          btnname="4"
          data-testid="num-li"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            4
          </button>
        </li>
        <li
          btnname="5"
          data-testid="num-li"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            5
          </button>
        </li>
        <li
          btnname="6"
          data-testid="num-li"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            6
          </button>
        </li>
        <li
          btnname="1"
          data-testid="num-li"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            1
          </button>
        </li>
        <li
          btnname="2"
          data-testid="num-li"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            2
          </button>
        </li>
        <li
          btnname="3"
          data-testid="num-li"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            3
          </button>
        </li>
        <li
          btnname="0"
          data-testid="num-li"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            0
          </button>
        </li>
        <li
          btnname="."
          data-testid="num-li"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            .
          </button>
        </li>
      </ul>
      <ul
        className="aritmetic"
      >
        <li
          data-testid="oper-li"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            รท
          </button>
        </li>
        <li
          data-testid="oper-li"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            x
          </button>
        </li>
        <li
          data-testid="oper-li"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            -
          </button>
        </li>
        <li
          data-testid="oper-li"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            +
          </button>
        </li>
        <li
          data-testid="oper-li"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            =
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
